src/qibo/metrics/metrics.py

```diff
@@ -0,0 +1,152 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+import scipy
+
+
+def trace_distance(state: np.ndarray, target: np.ndarray):
+    """Trace distance between two quantum states
+
+    ..math::
+        T(\\rho, \\sigma) \\coloneqq \\frac{1}{2} \\, ||\\rho - \\sigma||_{1}
+
+    where :math:`||\\cdot||_{1}` is the Schatten 1-norm.
+
+    Args:
+        state: state vector or density matrix.
+        target: state vector or density matrix.
+
+    Returns:
+        Trace distance between state :math:`\\rho` and target :math:`\\sigma`.
+
+    """
+
+    if state.shape != target.shape:
+        raise TypeError(
+            f"State has dims {state.shape} while target has dims {target.shape}."
+        )
+    elif len(state.shape) >= 3:
+        raise TypeError(
+            f"Both objects must have dims either (k,) or (k,l), but have dims {state.shape} and {target.shape}"
+        )
+
+    if len(state.shape) == 1:
+        state = np.outer(state.conj(), state)
+        target = np.outer(target.conj(), target)
```

Comment:
Not 100% sure but I think it is safer to use `np.conj` here:
```suggestion
        state = np.outer(np.conj(state), state)
        target = np.outer(np.conj(target), target)
```
and in other places.

The reason is that if you are using qibo to get the state, for example from circuit simulation, then depending on the backend the state may not be `np.ndarray`. 

With the current backends the state could also be cupy array (stored on GPU) or `tf.Tensor`. From my experience calling numpy methods on these two objects usually works, but calling it as an object method (eg. `state.conj()`) may not. So I would suggest to stick to the first option, but we can check this when we add tests.

Reply:
Thank you. just made the changes.